File: src/WN_RELATIONS.cpp
#include <iostream>
#include <fstream>
#include <regex>

using namespace std;

string* idiomas = new string[5] { "cat", "eus", "glg","por","spa"};

string do_replace( string const & in, string const & from, string const & to )
{
  return regex_replace( in, regex(from), to );
}

string codificacion_categorias(string c)
{
    if(c == "n") return "1";
    else if (c == "v") return "2";
    else if (c == "a") return "3";
    else if (c == "r") return "4";
    else return "-1";
}

int main()
{
    for(int idiomaIndex = 0; idiomaIndex < 5; idiomaIndex++)
    {
        fstream ficheroLectura;
        fstream ficheroEscritura_hyp;
        fstream ficheroEscritura_der;
        fstream ficheroEscritura_vgp;
        fstream ficheroEscritura_per;
        fstream ficheroEscritura_ant;
        fstream ficheroEscritura_sim;
        fstream ficheroEscritura_sub;
        fstream ficheroEscritura_xphyp;
        fstream ficheroEscritura_rel;
        fstream ficheroEscritura_cat;
        fstream ficheroEscritura_rgloss;
        fstream ficheroEscritura_rterm;
        fstream ficheroEscritura_uterm;
        fstream ficheroEscritura_mm;
        fstream ficheroEscritura_mp;
        fstream ficheroEscritura_sa;
        fstream ficheroEscritura_bis;
        fstream ficheroEscritura_cau;
        fstream ficheroEscritura_fuz;
        fstream ficheroEscritura_ml;
        fstream ficheroEscritura_mmof;
        fstream ficheroEscritura_hp;
        fstream ficheroEscritura_m;
        fstream ficheroEscritura_in;
        fstream ficheroEscritura_ina;
        fstream ficheroEscritura_ind;
        fstream ficheroEscritura_ini;
        fstream ficheroEscritura_inl;
        fstream ficheroEscritura_inp;
        fstream ficheroEscritura_isd;
        fstream ficheroEscritura_itd;
        fstream ficheroEscritura_xpfuz;
        fstream ficheroEscritura_xpant;
        fstream ficheroEscritura_xpsim;
        fstream ficheroEscritura_near;



        size_t i;
        string datos[9];
        cout << "Abriendo ficheros...\n";

        ficheroLectura.open("../mcr/"+idiomas[idiomaIndex]+"WN/wei_"+idiomas[idiomaIndex]+"-30_relation.tsv", ios::in);
        ficheroEscritura_hyp.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_hyp.pl", ios::out);
        ficheroEscritura_der.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_der.pl", ios::out);
        ficheroEscritura_vgp.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_vgp.pl", ios::out);
        ficheroEscritura_per.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_per.pl", ios::out);
        ficheroEscritura_ant.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_ant.pl", ios::out);
        ficheroEscritura_sim.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_sim.pl", ios::out);
        ficheroEscritura_sub.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_sub.pl", ios::out);
        ficheroEscritura_xphyp.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_xphyp.pl", ios::out);
        ficheroEscritura_rel.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_rel.pl", ios::out);
        ficheroEscritura_cat.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_cat.pl", ios::out);
        ficheroEscritura_rgloss.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_rgloss.pl", ios::out);
        ficheroEscritura_rterm.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_rterm.pl", ios::out);
        ficheroEscritura_uterm.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_uterm.pl", ios::out);
        ficheroEscritura_mm.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_mm.pl", ios::out);
        ficheroEscritura_mp.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_mp.pl", ios::out);
        ficheroEscritura_sa.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_sa.pl", ios::out);
        ficheroEscritura_bis.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_bis.pl", ios::out);
        ficheroEscritura_cau.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_cau.pl", ios::out);
        ficheroEscritura_fuz.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_fuz.pl", ios::out);
        ficheroEscritura_ml.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_ml.pl", ios::out);
        ficheroEscritura_mmof.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_mmof.pl", ios::out);
        ficheroEscritura_hp.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_hp.pl", ios::out);
        ficheroEscritura_m.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_m.pl", ios::out);
        ficheroEscritura_in.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_in.pl", ios::out);
        ficheroEscritura_ina.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_ina.pl", ios::out);
        ficheroEscritura_ind.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_ind.pl", ios::out);
        ficheroEscritura_ini.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_ini.pl", ios::out);
        ficheroEscritura_inl.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_inl.pl", ios::out);
        ficheroEscritura_inp.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_inp.pl", ios::out);
        ficheroEscritura_isd.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_isd.pl", ios::out);
        ficheroEscritura_itd.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_itd.pl", ios::out);
        ficheroEscritura_xpfuz.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_xpfuz.pl", ios::out);
        ficheroEscritura_xpant.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_xpant.pl", ios::out);
        ficheroEscritura_xpsim.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_xpsim.pl", ios::out);
        ficheroEscritura_near.open("../"+idiomas[idiomaIndex]+"/Prolog/wn_near.pl", ios::out);

        if(!ficheroLectura.is_open() || !ficheroEscritura_hyp.is_open() || !ficheroEscritura_der.is_open() || !ficheroEscritura_vgp.is_open() || !ficheroEscritura_per.is_open() || !ficheroEscritura_sub.is_open() || !ficheroEscritura_xphyp.is_open() || !ficheroEscritura_ant.is_open() || !ficheroEscritura_sim.is_open() || !ficheroEscritura_rel.is_open() || !ficheroEscritura_cat.is_open() || !ficheroEscritura_rgloss.is_open() || !ficheroEscritura_rterm.is_open() || !ficheroEscritura_uterm.is_open() || !ficheroEscritura_mm.is_open() || !ficheroEscritura_mp.is_open() || !ficheroEscritura_sa.is_open()|| !ficheroEscritura_bis.is_open() || !ficheroEscritura_cau.is_open() || !ficheroEscritura_fuz.is_open() || !ficheroEscritura_ml.is_open() || !ficheroEscritura_mmof.is_open() ||!ficheroEscritura_hp.is_open() || !ficheroEscritura_m.is_open() || !ficheroEscritura_in.is_open() || !ficheroEscritura_ina.is_open() || !ficheroEscritura_ind.is_open() || !ficheroEscritura_ini.is_open() || !ficheroEscritura_inl.is_open() || !ficheroEscritura_inp.is_open() || !ficheroEscritura_isd.is_open() || !ficheroEscritura_itd.is_open() || !ficheroEscritura_xpfuz.is_open() || !ficheroEscritura_xpant.is_open() || !ficheroEscritura_xpsim.is_open() || !ficheroEscritura_near.is_open())
        {
            cout << "No se han abierto correctamente\n";
            return 0;
        }

        cout << "Se han abierto correctamente.\n" << "Parseando a Prolog...\n";

        string line;

        while(getline(ficheroLectura, line))
        {
            i = 0;
            char *cadena = line.data();
            char *token =strtok(cadena, "\t");

            while(token != NULL)
            {
                datos[i] = token;
                token = strtok(NULL, "\t");
                i++;
            }

            if(datos[0] == "12")
            {
                ficheroEscritura_hyp << "hyp(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }else if(datos[0] == "33")
            {
                ficheroEscritura_ant << "ant(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ",0," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ",0)." << endl;
            }else if(datos[0] == "47")
            {
                ficheroEscritura_per << "per(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ",0," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ",0)." << endl;
            }else if(datos[0] == "19")
            {
                ficheroEscritura_sub << "sub(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }else if(datos[0] == "31")
            {
                ficheroEscritura_der << "der(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }else if(datos[0] == "52")
            {
                ficheroEscritura_vgp << "vgp(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ",0," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ",0)." << endl;
            }else if(datos[0] == "34")
            {
                ficheroEscritura_sim << "sim(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "21")
            {
                ficheroEscritura_xphyp << "xphyp(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }else if(datos[0] == "63")
            {
                ficheroEscritura_cat << "cat(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }else if(datos[0] == "64")
            {
                ficheroEscritura_rel << "rel(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "61")
            {
                ficheroEscritura_rgloss << "rgloss(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "66")
            {
                ficheroEscritura_rterm << "rterm(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "68")
            {
                ficheroEscritura_uterm << "uterm(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "7")
            {
                ficheroEscritura_mm << "mm(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "8")
            {
                ficheroEscritura_mp << "mp(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }else if(datos[0] == "49")
            {
                ficheroEscritura_sa << "sa(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ",0," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ",0)." << endl;
            }
            else if(datos[0] == "1")
            {
                ficheroEscritura_bis << "bis(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "2")
            {  
                ficheroEscritura_cau << "cau(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "3")
            {
                ficheroEscritura_fuz << "fuz(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "5")
            {
                ficheroEscritura_ml << "ml(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "6")
            {
                ficheroEscritura_mmof << "mmof(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "9")
            {
                ficheroEscritura_hp << "hp(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "10")
            {
                ficheroEscritura_m << "m(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "35")
            {
                ficheroEscritura_in << "in(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "36")
            {
                ficheroEscritura_ina << "ina(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "37")
            {
                ficheroEscritura_ind << "ind(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "38")
            {
                ficheroEscritura_ini << "ini(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "39")
            {
                ficheroEscritura_inl << "inl(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "40")
            {
                ficheroEscritura_inp << "inp(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "41")
            {
                ficheroEscritura_isd << "isd(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "42")
            {
                ficheroEscritura_itd << "itd(" << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << "," << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "44")
            {
                ficheroEscritura_xpfuz << "xpfuz(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "45")
            {
                ficheroEscritura_xpant << "xpant(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "46")
            {
                ficheroEscritura_xpsim << "xpsim(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
            else if(datos[0] == "60")
            {
                ficheroEscritura_near << "near(" << codificacion_categorias(datos[2]) << datos[1].substr(7,8) << "," << codificacion_categorias(datos[4]) << datos[3].substr(7,8) << ")." << endl;
            }
        }

        cout << "Idioma "+idiomas[idiomaIndex]+" parseado correctamente.\n";
        ficheroLectura.close();
        ficheroEscritura_hyp.close();
    }

    cout << "¡Proceso finalizado!";
    return 0;
}

File: src/WN_RELATIONS.py
import pandas as pd
import time

idiomas = ['cat', 'eng', 'eus', 'glg', 'por', 'spa']

relaciones = {
    '12':['hyp',1],'33':['ant',2],'31':['der',2],'52':['vgp',4],'47':['per',4],'34':['sim',3],'19':['sub',1],'21':['xphyp',1],'64':['rel',3],
    '63':['cat',1],'61':['rgloss',3],'66':['rterm',3],'68':['uterm',3],'7':['mm',3],'8':['mp',3],'49':['sa',4],'1':['bis',3],
    '2':['cau',3],'3':['fuz',3],'5':['ml',1],'6':['ms',3],'9':['hp',3],'10':['m',1],'35':['in',1],'36':['ina',1],
    '37':['ind',1],'38':['ini',1],'39':['inl',1],'40':['inp',1],'41':['isd',1],'42':['itd',1],'44':['xpfuz',3],'45':['xpant',3],
    '46':['xpsim',3],'60':['near',3]
    }

def codificacionCategoria(c):
    categorias = {
        'n':'1',
        'v':'2',
        'a':'3',
        'r':'4'
        }
    return categorias[c]

def crearSynsetID(synset):
    synset = codificacionCategoria(synset[-1])+synset[7:15]
    return int(synset)

def crearDF(key, df):
    dfAux = df[df['Rel_ID'] == int(key)]
    if not(dfAux.empty):
        if relaciones[key][1] == 1:
            dfAux = dfAux.drop(columns = ['Rel_ID'])
            dfAux = dfAux[['Synset_Target', 'Synset_Source']]
            dfAux.rename(columns = { 'Synset_Target': 'Synset Source', 'Synset_Source': 'Synset Target'}, inplace = True)

            dfAux = dfAux.sort_values(by=['Synset Source', 'Synset Target'])
            dfAux = dfAux.reset_index(drop = True)

            dfAux.to_csv(i+"\PrologCSV\wn_"+relaciones[key][0]+".csv")

            ficheroEscritura = open(i+"\Prolog\wn_"+relaciones[key][0]+".pl", "w", encoding='utf-8')
            for index in dfAux.index:
                ficheroEscritura.write(relaciones[key][0]+"("+str(dfAux["Synset Source"][index])+","+str(dfAux["Synset Target"][index])+").\n")
            ficheroEscritura.close()

        if relaciones[key][1] == 2:
            dfAux = dfAux.drop(columns = ['Rel_ID'])

            W_Num = [0] * len(dfAux['Synset_Target'])
            dfAux['W_Num Source'] = W_Num
            dfAux['W_Num Target'] = W_Num

            dfAux.rename(columns = { 'Synset_Target': 'Synset Source', 'Synset_Source': 'Synset Target'}, inplace = True)
            dfAux = dfAux[['Synset Source', 'W_Num Source', 'Synset Target', 'W_Num Target']]

            dfAux = dfAux.sort_values(by=['Synset Source', 'Synset Target'])
            dfAux = dfAux.reset_index(drop = True)

            dfAux.to_csv(i+"\PrologCSV\wn_"+relaciones[key][0]+".csv")

            ficheroEscritura = open(i+"\Prolog\wn_"+relaciones[key][0]+".pl", "w", encoding='utf-8')
            for index in dfAux.index:
                ficheroEscritura.write(relaciones[key][0]+"("+str(dfAux["Synset Source"][index])+",0,"+str(dfAux["Synset Target"][index])+",0).\n")
            ficheroEscritura.close()
            
        if relaciones[key][1] == 3:
            dfAux = dfAux.drop(columns = ['Rel_ID'])
            dfAux.rename(columns = { 'Synset_Source': 'Synset Source', 'Synset_Target': 'Synset Target'}, inplace = True)

            dfAux = dfAux.sort_values(by=['Synset Source', 'Synset Target'])
            dfAux = dfAux.reset_index(drop = True)

            dfAux.to_csv(i+"\PrologCSV\wn_"+relaciones[key][0]+".csv")
            
            ficheroEscritura = open(i+"\Prolog\wn_"+relaciones[key][0]+".pl", "w", encoding='utf-8')
            for index in dfAux.index:
                ficheroEscritura.write(relaciones[key][0]+"("+str(dfAux["Synset Source"][index])+","+str(dfAux["Synset Target"][index])+").\n")
            ficheroEscritura.close()

        if relaciones[key][1] == 4:
            dfAux = dfAux.drop(columns = ['Rel_ID'])

            W_Num = [0] * len(dfAux['Synset_Target'])
            dfAux['W_Num Source'] = W_Num
            dfAux['W_Num Target'] = W_Num

            dfAux.rename(columns = { 'Synset_Source': 'Synset Source', 'Synset_Target': 'Synset Target'}, inplace = True)
            dfAux = dfAux[['Synset Source', 'W_Num Source', 'Synset Target', 'W_Num Target']]

            dfAux = dfAux.sort_values(by=['Synset Source', 'Synset Target'])
            dfAux = dfAux.reset_index(drop = True)

            dfAux.to_csv(i+"\PrologCSV\wn_"+relaciones[key][0]+".csv")

            ficheroEscritura = open(i+"\Prolog\wn_"+relaciones[key][0]+".pl", "w", encoding='utf-8')
            for index in dfAux.index:
                ficheroEscritura.write(relaciones[key][0]+"("+str(dfAux["Synset Source"][index])+",0,"+str(dfAux["Synset Target"][index])+",0).\n")
            ficheroEscritura.close()

# 1: 2 datos del revés

# 2: 4 datos del revés

# 3: 2 datos del derecho

# 4: 4 datos del derecho

for i in idiomas:
    inicio = time.time()
   
    ruta = "mcrCSV//"+i+"WN\wei_"+i+"-30_relation.csv"
    print("Leyendo dataframe "+ ruta)
    
    df = pd.read_csv(ruta, index_col=[0])
    df = df.drop(columns=['S_PoS', 'T_PoS', 'Conf', 'Info'])

    df.columns = ['Rel_ID', 'Synset_Source', 'Synset_Target']

    df['Synset_Source'] = df['Synset_Source'].apply(crearSynsetID)
    df['Synset_Target'] = df['Synset_Target'].apply(crearSynsetID)

    keys = relaciones.keys()

    for key in keys:
        print('Se está traduciendo la relación: '+relaciones[key][0])
        crearDF(key, df)

    final = time.time()
    print('Proceso en ruta: ' + ruta + ' finalizado. Ha tardado '+str(final-inicio)+'.\n')
    